src/coreclr/src/tools/Common/Compiler/CompilerTypeSystemContext.cs

```diff
@@ -330,6 +331,68 @@ private PdbSymbolReader OpenAssociatedSymbolFile(string peFilePath, PEReader peR
 
             return reader;
         }
+
+        private object _normalizedLock = new object();
+        private HashSet<TypeDesc> _nonNormalizedTypes = new HashSet<TypeDesc>();
+        private Dictionary<TypeDesc, TypeFlags> _normalizedTypeCategory = new Dictionary<TypeDesc, TypeFlags>();
+
+        public TypeFlags NormalizedCategoryFor4ByteStructOnX86(TypeDesc type)
+        {
+            lock(_normalizedLock)
+            {
+                if (_nonNormalizedTypes.Contains(type))
+                    return type.Category;
+                
+                if (_normalizedTypeCategory.TryGetValue(type, out TypeFlags category))
+                    return category;
+
+                if (Target.Architecture != TargetArchitecture.X86)
+                {
+                    throw new NotSupportedException();
+                }
+
+                if ((type.Category != TypeFlags.ValueType) || (type.GetElementSize().AsInt != 4))
```

Comment:
This check is cheap. Maybe we only need to cache this for 4-byte valuetypes and we can get by without locking/caching otherwise? Might also be worth to frontload the I4/U4/I/U check so that we can do all the common things without taking the lock and doing hashtable lookups.

Reply:
I moved this check up. The rest might/might not be valuable. I'll wait for a profile for the rest.